Add lambda support to indentation check

pietern opened the pull request:
This is a first pass for lambda support in the indentation check. All unit and integration tests pass. The test case is based on the test case in this branch: https://github.com/checkstyle/checkstyle/compare/master...pirat9600q:IndentationCheck_issue281. Not all cases will be covered in this commit, but at least it's a start. I ran checkstyle against a codebase I'm working on and all the lambda related warnings disappear with this commit.

Also see #281.

mkordas commented:
Thanks for the great update, just couple of things:
* There are multiple builds failed at https://travis-ci.org/checkstyle/checkstyle/builds/73946529, just run `mvn clean verify` before commit
* put issue number to the commit message according to http://checkstyle.sourceforge.net/contributing.html
* Avoid mysterious abreviations in comments like`w.r.t.`
* please run analysis on project with many lambdas like infinispan/infinispan

pietern commented:
Thanks for the feedback. Addressed your comments. Only thing remaining is a cobertura coverage error.

pietern commented:
The coverage is now fixed by removing an apparently superfluous branch :)

I ran this against infinispan and it didn't crash (it had a zillion warnings/errors though).

Let me know if there is anything else that is blocking this from going in.

uschi2000 commented:
Hi folks,
In which release is this change expected to be available?
  Thanks! R.

romani commented:
Most likely at the end of September, we will not have time to review and test till end of August

romani commented:
@mkordas , can you finish code-review for this PR?

romani commented:
@pietern , sorry for delay , please rebase your PR over latest code we are about to start reviewing your code.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:
```diff
@@ -0,0 +1,67 @@
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+
+/**
+ * Handler for lambda statements.
```
more common is to say `lambda expressions` instead of `lambda statements`

pietern replied:
You're right, I started out with a copy of one of the other handlers.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:
```diff
@@ -0,0 +1,67 @@
+            return new IndentLevel(getLineStart(firstChild));
+        }
+
+        // The start of the lambda isn't the first element on the line;
```
Comment `The start of the lambda isn't the first element on the line` is contradicting with comment above: `If the start of the lambda isn't the first element on the line`. What about the case when lambda is the first element on the line?

pietern replied:
The second command should be negated, because it's when the lambda IS the first element.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java:
```diff
@@ -160,7 +160,8 @@ private static int getFirstNodeIndent(DetailAST node) {
 
-            if (curNode.getType() == TokenTypes.OBJBLOCK) {
+            if (curNode.getType() == TokenTypes.OBJBLOCK
+                    || curNode.getType() == TokenTypes.SLIST) {
```
Can you explain why do we need this change?

pietern replied:
It's needed because a lambda body is an SLIST, which needs to be captured by the line wrapper.

mkordas commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/indentation/InputLambda1.java:
```diff
@@ -0,0 +1,178 @@
+      x.run(); //indent:6 exp:6
+    }; //indent:4 exp:4
+
+    Runnable r1 = () -> { //indent:4 exp:4
```
we should also add more tests for more unusual cases like below:
```java
Runnable r7 = ()
     ->
    {
    }

Runnable r8 =
    ()
     ->
    {}

pietern replied:
Sounds good, I'll add a few more.

pietern commented:
Thank you for reviewing. Addressing the comments in a new commit.

mkordas commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/indentation/InputLambda1.java:
```diff
@@ -0,0 +1,178 @@
+          x.run(); //indent:10 exp:10
+        }); //indent:8 exp:8
+
+    SomeInterface i1 = (LongTypeName //indent:4 exp:4
```
I don't see any declaration of `SomeInterface` - please make input compilable in Java 8.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:
```diff
@@ -0,0 +1,67 @@
+     * @param parent        the parent handler
+     */
+    public LambdaHandler(IndentationCheck indentCheck,
+                         DetailAST ast, AbstractExpressionHandler parent) {
```
Aligning method parameters and Javadoc looks nice, but whenever any name or modifier will be changed, all the lines must be changed instead of one affected. I would go without any alignment.

mkordas commented:
Please use src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/coding/InputReturnCountLambda.java as a source of ideas for some more complicated lambdas.

For example these are valid lambdas:
```java
    Supplier<Supplier<Integer>> lambdas() {
        return (
        )
                ->
                (

                )
                        ->
                        1
                ;
```

mkordas commented:
@pietern, did you have time to work on this PR?

pietern commented:
@mkordas I incorporated your comments, added some more test cases where the lambda's argument list or the lambda token itself needs to be indented, and rebased on top of master.

mkordas commented:
Thanks @pietern, I'll look into it shortly.

mkordas commented:
@pietern 
> The statement-only lambda is not yet checked properly in this pull request, but can be addressed later in my opinion.

Please create separate issue for that.

Otherwise the changes look good for me. @romani, can you do the final review?

pietern commented:
@mkordas Sounds good, added #2184 for that.

romani commented:
@pietern, thank for your persistence to resolve that issue.

While I reviewing a code ,  ......

please generate reports by our checkstyle-tester tool - https://github.com/checkstyle/contribution/tree/master/checkstyle-tester , all instructions are in README .
you need to generate reports on checkstyle binaries before and after your fix. Nothing should be changed on projects which does not use LAMBDAs. One report pair should contain LAMBDAs to let you and us verify that your code works fine.

romani commented:
@pietern , please make a link to #281 at commit message.

romani commented:
@pietern , lambda projects - https://github.com/checkstyle/checkstyle/pull/2176#issuecomment-140888598

pietern commented:
Thanks for the hints, I'll check it out against those projects.

What do you want me to say here though? Just that the reports didn't change? Or do you also want a copy of those reports if/when I can generate them?

mkordas commented:
@pietern, if there is no difference, we don't need any reports. If there are differences, we need to analyze them, so any form convenient for you would be fine.

romani commented:
> r do you also want a copy of those reports if/when I can generate them?

jsut put them to your github.io to share with us. See DEPLOY at https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/README.md , example of report sharing - https://github.com/checkstyle/checkstyle/pull/2157#issue-106026354

pietern commented:
Sounds good, I'll give it another shot. Tried before, but not sure what you were looking for.

Can I limit the checks to just the indentation check to limit the size of these reports?

mkordas commented:
@pietern, just Indentation is enough and you can also tune configuration so it gives less errors on typical code.

romani commented:
release 6.11will be on this Sunday (27th of September), it would be good to finish this PR before release

pietern commented:
I rebased my lambda change on top of HEAD and got the following results:

 | before | after
-------|-------|----------
RxJava|811|811
infinispan|33604|31593
jOOL|859|1430
protonpack|60|23

The increase in warnings in jOOL can be attributed to formatting as found in [Tuple.java](https://github.com/jOOQ/jOOL/blob/master/src/main/java/org/jooq/lambda/tuple/Tuple.java). The handler expects these lambdas to follow line wrapping indentation (which I had set to 8 for the checks for this repository). With line wrapping set to 4, the number of warnings reduces to 454.

romani commented:
@pietern , thanks a lot but please share report at github.io to let us validate. All diff cases need to be reviwed.

pietern commented:
@romani Yep, working on it.

pietern commented:
The checks I used for the different projects are located here:
https://github.com/pietern/checkstyle-reports/tree/gh-pages/lambda-1548

The reports are located here:
https://pietern.github.io/checkstyle-reports/

mkordas commented:
@pietern, thanks for the reports. I would focus on protonpack results, as they are small. I see that your changes fixed some violations, but also added new ones. I took the code from [newly introduced violations](https://pietern.github.io/checkstyle-reports/lambda-1548/after/protonpack/site/xref/protonpack/src/main/java/com/codepoetics/protonpack/AggregatingSpliterator.html#L25) and IntelliJ says that code is properly formatted. Can you explain this?

romani commented:
I will try to review some part of "infinispan" . @pietern , please do detailed analysis of "jOOL".

pietern commented:
@mkordas Yes, it's because lines 25-30 are expected to be indented by "basicIndent", just like any other block. I can add 2x basicIndent as allowed indentation level as well, which should resolve this issue. Let me know what you think.

pietern commented:
@romani I'll take a closer look at jOOL.

mkordas commented:
@pietern, we should allow either both offsets or find some common rules in IDEs. When to use (in this case) 4 or 8 spaces.

pietern commented:
@mkordas I checked that file in IntelliJ and it appears that it is indented with the base indent + continuation indent. This is because the lambda is actually an argument to a function, which must be what's triggering the continuation indentation. The lambda check in this PR doesn't take that into account and only looks at the indentation of the line start. I can change it to allow either that as base level OR that + continuation indent as base level.

mkordas commented:
@pietern, yes, let's add support for this + test cases. False-positives are not allowed :)

romani commented:
ok , I will wait for next fix and report regeneration.

pietern commented:
@mkordas I found out that what triggers the continuation indent is the closing parenthesis being on a separate line. If I nest the expression one level deeper (through a call to an identity function), the continuation indentation is applied another time. I'll try to incorporate this in my change.

pietern commented:
This is the diff between the previous commit and the new one: https://gist.github.com/pietern/6f80967e15931c7b8d8e

Adds support for the continuation indentation as present in protonpack and as executed by IntelliJ.

pietern commented:
Haven't had time to rerun the reports yet, doing that later tonight.

pietern commented:
New results:

 |before|after
------|------|-----
RxJava|811|869
infinispan|33604|31637
jOOL|454|454
protonpack|60|25

The new protonpack violations are caused by inconsistent indenting. The project uses an 8 space continuation but indents the lambda in AggregatingSpliterator.java with 4 spaces, where 8 spaces should have been used (you can verify by auto indenting using IntelliJ).

Didn't investigate the new RxJava violations yet.

pietern commented:
Also updated the gist above with the new diff.

romani commented:
infinispan diff - https://www.diffchecker.com/sqnddfdz

romani commented:
during diff investigation please be attentive where in "after" report there is change or addition of new violation. Each found false positive should be added as test case.

false positives in "after":
1)
infinispan/core/src/main/java/org/infinispan/interceptors/distribution/DistributionBulkInterceptor.java
Severity	Category	Rule	Message	Line
 Warning	indentation	Indentation	'lambda arguments' have incorrect indentation level 17, expected level should be 15.	135

https://pietern.github.io/checkstyle-reports/lambda-1548/after/infinispan/site/xref/infinispan/core/src/main/java/org/infinispan/interceptors/distribution/DistributionBulkInterceptor.html#L135

2)
infinispan/core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java
 Warning	indentation	Indentation	'method def modifier' have incorrect indentation level 18, expected level should be one of the following: 21, 24, 27.	406

3)
infinispan/core/src/main/java/org/infinispan/stream/impl/AbstractCacheStream.java
Severity	Category	Rule	Message	Line
 Warning	indentation	Indentation	'lambda arguments' have incorrect indentation level 14, expected level should be 12.	354

mkordas commented:
@pietern 
TeamCity noticed allocation of zero length array and unnecessary 'this' qualifiers - please fix it.

pietern commented:
Working on the infinispan test cases.

romani commented:
@pietern and @mkordas ,

One thing we should clearly understand: I am completely ok to merge and release non complete lambda support, but I am NOT ok if we introduce violations on code that does not have lambda at all (majority of projects still use jdk6, jdk7).

incomplete Lambda support:
- it is better then nothing, user will benefit from it, even it is not complete.
- we do not close issue, and continue working with lambda false-positives, it is ok to fix issue in few steps(few PRs)
- merging first PR will help to validate all other PRs and changes before and after should be minimal.

So lets focus on non-lambda false positives ! lets test project that do not have lambda at all, and if reports(before/after) are the same we are ok to merge.

pietern commented:
@romani Sounds good to me :)

Still looking why the infinispan method def modifier triggers a warning. The others can be attributed to inconsistent line continuation indenting in that repository; sometimes they use 6, other times they use 8. I used 6 in my configuration, so this triggers a warning for 8. Notice that since this is a new check, it will always produce new warnings of this type (lambda arguments).

I pushed a new change with the inspections fixed.

pietern commented:
@romani With the latest change we should be good on not breaking existing code. I expanded the set of allowed indentations for children to include line wrapping (which is what was originally intended, over basic indenting, it looks like) and line wrapping based on the parent's suggested level if the right paren is the first on the line. Both additions are tested in the lambda test input files (because it seems the lambda is one of the few that checks the indentation of the arguments, which applies in method call arguments as well). This addresses some of the lambda false positives that @mkordas mentioned for protonpack.

For the method call handler diff, see: https://github.com/checkstyle/checkstyle/commit/2224d1e694a4edb75888e43b378d795fe194619d#diff-77dfb9ce50cfab374af2799f71e68cfb

I pushed the new reports and this is the summarized result:

 |before|after
----|----|----
RxJava|811|807
infinispan|33604|31617
jOOL|454|454
protonpack|60|23

pietern commented:
Regarding the infinispan method def warning: these were caused by restricting the set of allowed indentations for children in the method call handler. I updated this to allow *more* than it allowed before, so it shouldn't happen anymore.

romani commented:
@pietern , please generate reports(before/after) on projects that do not have lambda's (hibernate, jdk7, spring) and confirm that there are no difference in reports.

@mkordas , please confirm that violations for RxJava and protonpack are valid.

I will review 'infinispan'.

As all of us confirm that results are good - I will merge this PR (I can even trigger release 6.11.1 as to speed up testing of this by community).

All false positives in Lambda that we still have should be posted as comments to issue, issue is not going to be closed for now.

pietern commented:
@romani Sounds good. Is the broadened set of acceptable indentations acceptable to you? I expect this to be the only material difference in the reports before/after this change, since it applies to all method calls.

romani commented:
Please give me example, i do not understand you

mkordas commented:
@romani I confirm that both protonpack and RxJava reports look really well. I haven't noticed anything suspicious. All changes are expected and beneficial for the check.

romani commented:
I updated my https://github.com/checkstyle/checkstyle/pull/1548#issuecomment-143462332 with list of false-positives at Lambda support.  I did not noticed any other serious changes that could block merge.

@pietern , please confirm your part of validation.

pietern commented:
@romani Re: the broadened set of acceptable indentations: that `suggestedChildLevel` returns more acceptable indentation levels than before. It is not just a single level that it returns.

Some more results for this change:

 |before|after
----|----|----
hibernate-orm|68604|68604
spring-framework|18586|18515

The fewer number of warnings in spring is caused by a few of lambdas in the code base that are now correctly checked. I checked the diff for hibernate and the warnings are identical. The only difference is that there is one method call with indented arguments where more indentation levels are acceptable, if you look at the diff of `checkstyle.txt` for hibernate (per the `suggestedChildLevel` comment above). I pushed the results for both projects (before and after) to the same repository. The directories contain `checkstyle.txt` for easy diff'ing (see https://github.com/pietern/checkstyle-reports/blob/gh-pages/gen-txt.sh).

pietern commented:
@romani I reviewed the false positives you listed for infinispan, but the first three of them are actually real positives; the code base has mixed use of 6 and 8 characters for line continuation. I ran the test with 6 (given that the basic offset is 3), which explains a few of the false positives for cases where 8 is used.

romani commented:
@pietern , please copy paste each false-positive and by trailing comments describe how much spaces code have and what is expected and what is configuration. Any general talk about indentation will never succeeded.

pietern commented:
Configuration for infinispan: http://pietern.github.io/checkstyle-reports/lambda-1548/check_infinispan.xml

1) [DistributionBulkInterceptor.html#L135](https://pietern.github.io/checkstyle-reports/lambda-1548/after/infinispan/site/xref/infinispan/core/src/main/java/org/infinispan/interceptors/distribution/DistributionBulkInterceptor.html#L135) -- indentation is 8, expected: 6. The other method call arguments are not checked at all, only the lambda is checked for correct indentation.

2) [AbstractCacheStream.html#L354](https://pietern.github.io/checkstyle-reports/lambda-1548/after/infinispan/site/xref/infinispan/core/src/main/java/org/infinispan/stream/impl/AbstractCacheStream.html#L354) -- indentation is 8, expected: 6. Same as (1).

3) [DistributedDoubleCacheStream.html#L150](https://pietern.github.io/checkstyle-reports/lambda-1548/after/infinispan/site/xref/infinispan/core/src/main/java/org/infinispan/stream/impl/DistributedDoubleCacheStream.html#L150) -- indentation is 8, expected: 6. Same as (1).

4) [FunctionalListenerAssertions.html#L254](https://pietern.github.io/checkstyle-reports/lambda-1548/after/infinispan/site/xref/infinispan/core/src/test/java/org/infinispan/functional/FunctionalListenerAssertions.html#L254) -- agree this is a false positive. This is caused by the behavior expected in [InputLambda1.java#L205](https://github.com/checkstyle/checkstyle/pull/1548/files#diff-07a0c5902c0235123d0ed0f7e54c9ae2R205), where the right paren on the new line triggers an expectation of the child being indented with a line continuation amount (being 6), whereas the indentation here is equal to the basic offset (being 3).

romani commented:
merged as FF

romani commented:
not finished discussion is moved to #281

pietern commented:
Great news, thanks @romani !
